drivers/flash/flash_mspi_nor_mx.c

```diff
@@ -0,0 +1,989 @@
+	struct mspi_scramble_cfg scramble_cfg;
+	mspi_timing_cfg timing_cfg;
+	struct mspi_xfer trans;
+	struct mspi_xfer_packet packet;
+
+	struct k_sem lock;
+	uint8_t jedec_id[JESD216_READ_ID_LEN];
+};
+
+static bool flash_mspi_nor_mx_address_is_valid(const struct device *dev, off_t addr, size_t size)
+{
+	const struct flash_mspi_nor_mx_config *dev_cfg = dev->config;
+	uint32_t flash_size = dev_cfg->mem_size;
+
+	return (addr >= 0) && ((uint64_t)addr + (uint64_t)size <= flash_size);
+}
+
+/* Command to the flash for writing parameters */
+static int flash_mspi_nor_mx_command_write(const struct device *flash, uint8_t cmd, uint32_t addr,
+					   uint16_t addr_len, uint32_t tx_dummy, uint8_t *wdata,
+					   uint32_t length)
+{
+	const struct flash_mspi_nor_mx_config *cfg = flash->config;
+	struct flash_mspi_nor_mx_data *data = flash->data;
+	int ret;
+
+	data->packet.dir = MSPI_TX;
+	data->packet.cmd = cmd;
+	data->packet.address = addr;
+	data->packet.data_buf = wdata;
+	data->packet.num_bytes = length;
+
+	data->trans.async = false;        /* meaning : timeout mode */
+	data->trans.xfer_mode = MSPI_PIO; /* command_write is always in PIO mode */
+	data->trans.tx_dummy = tx_dummy;
+	data->trans.cmd_length = 1;
+	data->trans.addr_length = addr_len;
+	data->trans.hold_ce = false;
+	data->trans.packets = &data->packet;
+	data->trans.num_packet = 1;
+	data->trans.timeout = 10;
+
+	ret = mspi_transceive(cfg->bus, &cfg->dev_id, (const struct mspi_xfer *)&data->trans);
+	if (ret) {
+		LOG_ERR("MSPI Tx transaction failed with code: %d", ret);
+		return -EIO;
+	}
+
+	LOG_DBG("MSPI Tx transaction (cmd = 0x%x)", data->packet.cmd);
+
+	return ret;
+}
+
+/* Command to the flash for reading parameters */
+static int flash_mspi_nor_mx_command_read(const struct device *flash, uint8_t cmd, uint32_t addr,
+					  uint16_t addr_len, uint32_t rx_dummy, uint8_t *rdata,
+					  uint32_t length)
+{
+	const struct flash_mspi_nor_mx_config *cfg = flash->config;
+	struct flash_mspi_nor_mx_data *data = flash->data;
+	int ret;
+
+	data->packet.dir = MSPI_RX;
+	data->packet.cmd = cmd;
+	data->packet.address = addr;
+	data->packet.data_buf = rdata;
+	data->packet.num_bytes = length;
+
+	data->trans.async = false;        /* meaning : timeout mode */
+	data->trans.xfer_mode = MSPI_PIO; /* command_read is always in PIO mode */
+	data->trans.rx_dummy = rx_dummy;
+	data->trans.cmd_length = 1;
+	data->trans.addr_length = addr_len;
+	data->trans.hold_ce = false;
+	data->trans.packets = &data->packet;
+	data->trans.num_packet = 1;
+	data->trans.timeout = 10;
+
+	ret = mspi_transceive(cfg->bus, &cfg->dev_id, (const struct mspi_xfer *)&data->trans);
+	if (ret) {
+		LOG_ERR("MSPI Rx transaction failed with code: %d", ret);
+		return -EIO;
+	}
+
+	LOG_DBG("MSPI Rx transaction (cmd = 0x%x)", data->packet.cmd);
+
+	return ret;
+}
+
+/* Command to the flash for reading status register  */
+static int flash_mspi_nor_mx_status_read(const struct device *flash, uint8_t status)
```

Comment:
So the user context way did not work out?
The methods(auto polling) introduced in this function is not a common one thus should not appear like this in a shared driver. Others(controllers) will not be able to use this driver.
Also I'm confused with the use of async PIO but does not register callback?

Reply:
actually, I did not find a proper way to handle that with the user context

Reply:
I have a example raw implementation of autopoll from Ambiq controller. We don't support autopoll natively by hardware but the upper layer should suffice to demonstrate my suggestion.
Please checkout https://github.com/AmbiqMicro/ambiqzephyr/commits/apollo3p-mspi-autopoll-trial/ and this commit [39e35c83590bdef7096e4e13e53a35346cbfaa23](https://github.com/AmbiqMicro/ambiqzephyr/commit/39e35c83590bdef7096e4e13e53a35346cbfaa23)

Reply:
> ur own user context for autopolling in the device driver as this may

Ok, I can see the data->autopoll to pass to the mspi driver 
What is ye not really clear is
-  the goal of the magic
- the role of flash_mspi_atxp032_autopoll.match_ap/mask_ap compared to mspi_ambiq_autopoll_cfg->matc flash_mspi_atxp032_autopoll-> mask)
- in the mspi_stm32_transceive I still have a check on the command to send (RDSR) but I do not see where the autopoll_cb is used

Reply:
> the goal of the magic

It is for type safety to make sure the upper layer did pass the correct structure one defined as the `ctx` defined in `struct mspi_callback_context` is of type void *.

> the role of flash_mspi_atxp032_autopoll.match_ap/mask_ap compared to mspi_ambiq_autopoll_cfg->match
/ flash_mspi_atxp032_autopoll-> mask

I try to design the autopoll feature to not just about pulling one byte and verify one bit in the flash status register. So in `mspi_ambiq_autopoll_cfg` the `mask` and `match` are array of values and masks that one can use to match with what will be read from the MSPI interface. You don't have to follow through on this.

> in the mspi_stm32_transceive I still have a check on the command to send (RDSR) but I do not see where the autopoll_cb is used

[Here](https://github.com/FRASTM/zephyr/blob/4bbbe0238d296c3e44e76333d5e1df876953baf6/drivers/mspi/mspi_stm32.c#L399) everything is configured and waiting for a interrupt, correct? What should happen is that your `HAL_XSPI_IRQHandler` should call the autopull_cb when the interrupt fires. I see you have copied `mspi_autopoll_callback` to your file, this is equivalent to your HAL_XSPI_xxxxCallback. Without looking into your hal, you may need to find a way to register `mspi_autopoll_callback` to your hal. Similarly to how `mspi_dma_callback` is being registered.

Below is irrelevent to autopoll
`MSPI_DMA : async/sync is meaningless with DMA (no DMA IT function)`
I'm curious that since you have `mspi_dma_callback` this means dma operation can be async, why are you saying in your comment that it is meaningless.